### PR TITLE
Implement basic build infrastructure and CI testing

### PR DESCRIPTION
**Pull request by duytnguyendtn**
This PR cleans up the initial code pushed up in the repo and introduces various infrastructure requirements. To highlight:

- "pip install"-able
- Testable via pytest
- PEP8 compliance
- GitHub Actions Workflow files
- Python Packaging Manifest file
- Minimal pyproject.toml

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[project]
+name = "pypolyclip"
+description = "A python driver for polyclip written in C by JD Smith."
+dynamic = [
+    "version"
+]
+
+[build-system]
+requires = ["setuptools", "wheel", "numpy"]
```
Comment (larrybradley) [inline pyproject.toml:9]:
```suggestion
requires = ["setuptools", 
            "setuptool_scm>=6.2",
            "oldest-supported-numpy"]
build-backend = "setuptools.build_meta"
```

@pllim Should `oldest-supported-numpy` still be used now that `numpy 1.25` is released?

File: pypolyclip/pypolyclip.py [hunk 2]
```diff
@@ -126,87 +134,90 @@ def single(x, y, nxy):
     ----------
     x : int or float
         The x coordinates of the polygon corners
-    
+
     y : int or float
         The y coordinates of the polygon corners
 
     nxy : list, tuple, or `np.ndarray`
         The size of the pixel grid.
```
Comment (larrybradley) [inline pypolyclip/pypolyclip.py:142]:
```suggestion
           The size of the pixel grid.
```
The description lines should line up with line above.  This is an issue in several places.

File: pypolyclip/pypolyclip.py [hunk 3]
```diff
@@ -126,87 +134,90 @@ def single(x, y, nxy):
...
+        np.asarray(x, dtype=FLT),
+        np.asarray(y, dtype=FLT),
+        nverts,
+        px_out,
+        py_out,
+        inds,
+        nclip,
+        areas,
+        ri_out,
+    )
```
Comment (larrybradley) [inline pypolyclip/pypolyclip.py:206]: I really dislike `black` style formatting.  IMHO, the original was better, but this is a stylistic issue.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+[project]
+name = "pypolyclip"
+description = "A python driver for polyclip written in C by JD Smith."
+dynamic = [
+    "version"
+]
```
Comment (larrybradley) [inline pyproject.toml:6]: This project information is duplicated in `setup.cfg`.  Ideally, it should go in one or the other.

File: .github/workflows/ci_cron_weekly.yml
Comment (larrybradley) [inline .github/workflows/ci_cron_weekly.yml:1]: This should be re-written to use a job maxtrix (like `ci_workflows.yml`) to avoid repeated code.

File: .github/workflows/ci_workflows.yml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      with:
+        python-version: ${{ matrix.python }}
+    - name: Install base dependencies
+      run: python -m pip install --upgrade pip tox
+    - name: Test/run with tox
+      run: tox -e ${{ matrix.toxenv }} -- ${{ matrix.toxposargs }}
+    - name: Upload coverage to codecov
+      if: "contains(matrix.toxenv, '-cov')"
+      uses: codecov/codecov-action@v3
+        env: CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
```
Comment (larrybradley) [inline .github/workflows/ci_workflows.yml:91]: This isn't needed in the GitHub repo is public.

**Review by larrybradley — CHANGES_REQUESTED** [review]
Thanks, @duytnguyendtn!  I've left several comments.  Please let me know if you have questions.

File: MANIFEST.in [hunk 6]
```diff
@@ -0,0 +1,9 @@
+include README.rst
+include CHANGES.rst
+include LICENSE.rst
```
[the remark this answers is not in the document]
Comment (larrybradley) [reply MANIFEST.in:3]: I see that the AURA license is already in the `LICENSE` file (appended at the end).  I think that's probably fine.

File: MANIFEST.in [hunk 7]
```diff
@@ -0,0 +1,9 @@
+include README.rst
+include CHANGES.rst
```
[the remark this answers is not in the document]
Comment (duytnguyendtn) [reply MANIFEST.in:2]: Added in https://github.com/spacetelescope/pypolyclip/pull/4/commits/25d2bfa992328e184bc3f102369bf7e9f94b1ad9!

File: MANIFEST.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (duytnguyendtn) [reply MANIFEST.in]: Closing since AURA License is already included

[the remark this answers is not in the document]
Comment (duytnguyendtn) [reply MANIFEST.in]: Renamed the license file in https://github.com/spacetelescope/pypolyclip/pull/4/commits/25d2bfa992328e184bc3f102369bf7e9f94b1ad9

File: pypolyclip/_version.py
[the remark this answers is not in the document]
Comment (duytnguyendtn) [reply pypolyclip/_version.py:1]: .gitignore updated and version file removed in https://github.com/spacetelescope/pypolyclip/pull/4/commits/25d2bfa992328e184bc3f102369bf7e9f94b1ad9

File: pypolyclip/tests/test_pypolyclip.py
[the remark this answers is not in the document]
Comment (duytnguyendtn) [reply pypolyclip/tests/test_pypolyclip.py:1]: Done in https://github.com/spacetelescope/pypolyclip/pull/4/commits/25d2bfa992328e184bc3f102369bf7e9f94b1ad9

File: pyproject.toml (hunk 4 above)
Comment (duytnguyendtn) [reply pyproject.toml]: Stripped duplicated info in https://github.com/spacetelescope/pypolyclip/pull/4/commits/25d2bfa992328e184bc3f102369bf7e9f94b1ad9

File: .github/workflows/ci_workflows.yml (hunk 5 above)
Comment (duytnguyendtn) [reply .github/workflows/ci_workflows.yml]: Done in https://github.com/spacetelescope/pypolyclip/pull/4/commits/6a94d100042b02d859529fe68e213165065efca6

**Comment by pllim** [comment]
Did someone ping me here? I don't see it but I also don't know why I am getting notified of this PR.

File: pyproject.toml (hunk 1 above)
Comment (larrybradley) [reply pyproject.toml:9]: @pllim I pinged you here.

Comment (pllim) [reply pyproject.toml:9]: Ah, thanks! Must have been hidden by all the other comments, so didn't show up in quick search. 

You could drop it but there are some unresolved discussions about ABI (see `seberg` comments) and `#define` over at `astropy`. See if they apply to you or not first: https://github.com/astropy/astropy/pull/14949

Comment (larrybradley) [reply pyproject.toml:9]: I suppose we should just use `oldest-supported-numpy` for now.  We can revisit this later once things are settled.

File: .github/workflows/ci_cron_weekly.yml
Comment (larrybradley) [reply .github/workflows/ci_cron_weekly.yml:1]: @duytnguyendtn You should be able to follow the pattern in the `ci_workflows.yml` file for defining the test matrix.

**Comment by duytnguyendtn** [comment]
@larrybradley Pushed up that missing tox file in https://github.com/spacetelescope/pypolyclip/pull/4/commits/abe4161936604a2a6e12d9d10be48da98ca56e61. Largely just borrowed the one I added in slitlessutils. Taking care of https://github.com/spacetelescope/pypolyclip/pull/4#discussion_r1260033025 now!

File: .github/workflows/ci_cron_weekly.yml
Comment (duytnguyendtn) [reply .github/workflows/ci_cron_weekly.yml:1]: Done in https://github.com/spacetelescope/pypolyclip/pull/4/commits/9f3d3dd3fc4efb65cbcfe27f2b31965037697fd4! I just migrated the single cron test I had in there to a test matrix; let me know if you think we need to add additional ones. We don't have any other dependencies here other than numpy as a build dep, but I can add devdeps if I missed something.

Thanks!

**Comment by duytnguyendtn** [comment]
Ok I think I got everything; ready for re-review!

File: .github/workflows/ci_cron_weekly.yml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    if: github.repository == 'spacetelescope/pypolyclip'
+    matrix:
+      include:
+        # NOTE: we don't have devdeps here
+        - name: Python 3.11
+          os: ubuntu-latest
+          python: 3.11
+          toxenv: py311-test
+          toxposargs: --run-slow
+
```
Comment (larrybradley) [inline .github/workflows/ci_cron_weekly.yml:44]:
```suggestion
        - name: linkcheck
          os: ubuntu-latest
          python: 3.11
          toxenv: py311-linkcheck
```

File: tox.ini [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+requires =
+    setuptools >= 30.3.0
+    pip >= 19.3.1
+isolated_build = true
+
+[testenv]
+# Suppress display of matplotlib plots generated during docs build
+setenv =
+    MPLBACKEND=agg
+    JUPYTER_PLATFORM_DIRS=1
```
Comment (larrybradley) [inline tox.ini:17]:
```suggestion
```
Why is this line here?

**Review by larrybradley — CHANGES_REQUESTED** [review]
Please let me know if you have questions.

File: tox.ini [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    predeps: true
+    !predeps: false
+
+[testenv:linkcheck]
+changedir = docs
+description = check the links in the HTML docs
+extras = docs
+commands =
+    pip freeze
+    sphinx-build -W -b linkcheck . _build/html
```
Comment (larrybradley) [inline tox.ini:63]:
```suggestion
```

File: tox.ini [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+description = check code style, e.g. with flake8
+deps = flake8
+commands = flake8 pypolyclip --count
+
+[testenv:securityaudit]
+skip_install = true
+changedir = .
+description = security audit with bandit
+deps = bandit
+commands = bandit -r pypolyclip -c .bandit.yaml
```
Comment (larrybradley) [inline tox.ini:77]: This test will fail because there is no `.bandit.yaml` file in the repo.  Please add one.

**Comment by larrybradley** [comment]
Thanks, @duytnguyendtn .  The `test`, `codestyle`, `pep517`, `securityaudit` tests all pass locally now. 

`linkcheck` still needs to be removed the `tox.ini` `envlist`.

**Comment by larrybradley** [comment]
And this PR has grown to 19 commits already.  Can you clean up the commit history by squashing some of those commits?

**Comment by duytnguyendtn** [comment]
@larrybradley this line, right? https://github.com/spacetelescope/pypolyclip/pull/4/commits/58140998351b7a17265c8966f576f9c9461489ec

Yes, I'll definitely squash and merge this PR to collapse the commits

**Review by larrybradley — APPROVED** [review]
Thanks, @duytnguyendtn
